Add pathway analysis muddiest points notebook (#37)
AlexsLemonade/2021-june-training: supplemental_notebooks/2021-07-01_muddiest_points.Rmd

- [Saving GSEA plots for all significant pathways](#saving-gsea-plots-for-all-significant-results)
- [GSEA significance](#a-little-more-on-GSEA-significance)

### How to use this notebook: 

While it is fine to just read this page as is, if you want to get the Rmarkdown notebook itself to work with, you can follow the following steps:

1. Click the `Code` button in the upper right of this page and choose `Download Rmd` to save this notebook (`2021-07-01_muddiest_points.Rmd`) to your computer.
If you have RStudio installed on your computer, you can then open the file directly.

Otherwise you can add it to the CCDL RStudio server to work with it there with the following additional steps:

2. Navigate to RStudio server.  
3. In the `File` pane, navigate to `training-modules/scRNA-seq`.     
4. Click the `Upload` button in the `File` pane.   
5. Click `Choose file` and find this `Rmd` file on your computer, then click `OK`  

Now you can open up this file and run it on the Server. 

## Libraries

```{r}
library(magrittr)
# Package to run GSEA
library(clusterProfiler)
# Package that contains the MSigDB gene sets in tidy format
library(msigdbr)
```

> 

Reviewer: Adding a chunk to install ggridges
~~~suggestion

For some later plots, we are going to use the `ggridges` package, which may not be installed, so let's install it here (only if it is not already installed):

```{r}
if (! "ggridges" %in% installed.packages()){
  install.package("ggridges")
}
```
~~~

Author: I guess I must have installed it myself previously!

                             category = "C7")
```

### Run GSEA

Run GSEA using the same parameters as we did during instruction.

```{r}
gsea_results <- GSEA(geneList = lfc_vector,  # ordered ranked gene list
                     minGSSize = 25,  # minimum gene set size
                     maxGSSize = 500,  # maximum gene set set
                     pvalueCutoff = 0.05,
                     pAdjustMethod = "BH",  # correction for multiple hypothesis testing
                     TERM2GENE = dplyr::select(hs_immunologic_df,
                                               gs_name,
                                               gene_symbol))
```
Let's get the gene set names for all of the significant gene sets.

```{r}
significant_gene_sets <- gsea_results@result %>%  # tabular results
  dplyr::filter(p.adjust < 0.05) %>%  # filter based on adjusted p-value
  dplyr::pull(ID)  # pull the gene set names as a vector
```

### Save a GSEA plot for every significant 

First, let's make sure we have a `plots/hodgkins/gsea` directory to hold our output.

```{r}
plots_dir <- file.path("plots", "hodgkins", "gsea")
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

> First, let's look at (one) way to save each of the significant pathways using `purrr`.

Reviewer: Only get to be first once!
```suggestion
Now, let's look at (one) way to save each of the significant pathways using `purrr`.
```

```{r}
# Create a list of plots, where each element of the list will be 
plot_list <- purrr::map(
  as.list(significant_gene_sets),   # Make a list of significant gene sets
  ~ enrichplot::gseaplot(gsea_results,
                         geneSetID = .x,  # Pathway name
                         title = .x # Pathway name
  )
)

# Now we need to "walk" through both the plots and the gene set names

> plot_list %>%
>   purrr::walk2(as.list(significant_gene_sets),  # contains pathway names
>                ~ ggplot2::ggsave(plot = .x,   # plot itself (first list)
>                                  filename = file.path(plots_dir, 
>                                                       paste0(.y, # pathway name (second list)
>                                                              "_gsea_plot.png")),
>                                  width = 11, height = 7
>                ))

Reviewer: ```suggestion
purrr::walk2(
  plot_list, # the list of plots
  as.list(significant_gene_sets),  # the list of pathway names
  ~ ggplot2::ggsave( # the function we are applying
      plot = .x,   # plot itself (from the first list)
      filename = file.path(plots_dir, 
                           paste0(.y, # pathway name (from the second list)
                                  "_gsea_plot.png")),
      width = 11, 
      height = 7
    )
)
```

```

Here's how to do the same thing by looping through the vector of the significant pathway names with a `for` loop!

```{r}
for (pathway in significant_gene_sets) {
  # Create the plot and assign it to `p`
  p <- enrichplot::gseaplot(gsea_results,
                            geneSetID = pathway,  # Pathway name
                            title = pathway) # Pathway name
  # Save the plot, using the pathway name as part of the filename
  # Note that this will write over the files we saved with the purrr method!

>   ggplot2::ggsave(plot = p, filename = file.path(plots_dir,
>                                                  paste0(pathway, "_gsea_plot.png")),
>                   width = 11, height = 7)

Reviewer: Is this reformat necessary? No, but I think I like it better.
```suggestion
  ggplot2::ggsave(plot = p, 
                  filename = file.path(plots_dir, paste0(pathway, "_gsea_plot.png")),
                  width = 11, height = 7)
```

                     minGSSize = 15,  # minimum gene set size
                     maxGSSize = 500,  # maximum gene set set
                     pvalueCutoff = 1,  # Return everything!
                     pAdjustMethod = "BH",  # correction for multiple hypothesis testing
                     TERM2GENE = random_gene_set_df)  # data frame with randomly generated gene sets
```
Let's get the enrichment scores and normalized enrichment scores in a data frame we'll use for plotting.

```{r}
random_es <- gsea_random@result %>%
  # Just the pathway name, enrichment score, and normalized enrichment score
  dplyr::select(ID, enrichmentScore, NES) %>%
  # Create a new column that includes the gene set size as a factor in ascending
  # order
  dplyr::mutate(pathway_size = factor(
    stringr::str_remove(stringr::str_remove(ID ,"\\..*"), ".*\\_" ),
    levels = c(15, 25, 50, 100, 250, 500)
  ))

# Take a peek
head(random_es)
```

Let's create a ridgeline plot of enrichment scores.

```{r}
random_es %>%
  ggplot2::ggplot(ggplot2::aes(enrichmentScore,
                               pathway_size,
                               fill = pathway_size)) +

>   ggridges::geom_density_ridges() +

Reviewer: We don't have `ggridges`! We can show how to install it here (or a bit earlier) if we like.
